02-peer-protocol.md

```diff
@@ -50,10 +67,361 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
+
+#### *initiator* only
+
+A, *initiator* has two inputs and an output (the funding output).
+B, the *non-initiator* has nothing to contribute.
+n
+
+        +-------+                       +-------+
+        |       |--(1)- tx_add_input -->|       |
+        |       |<-(2)- tx_complete ----|       |
+        |       |--(3)- tx_add_input -->|       |
+        |   A   |<-(4)- tx_complete ----|   B   |
+        |       |--(5)- tx_add_output ->|       |
+        |       |                       |       |
+        |       |--(6)- tx_complete --->|       |
+        |       |<-(7)- tx_complete ----|       |
+        +-------+                       +-------+
+
+#### *initiator* and *non-initiator*
+
+A the *initiator* contributes 2 inputs and an output that they
+then remove.  B, the *non-initiator*, contributes 1 input and an output,
+but waits until A adds a second input before contributing.
+
+Note that if A does not send a second input, the negotiation will end without
+B's contributions.
+
+        +-------+                       +-------+
+        |       |--(1)- tx_add_input -->|       |
+        |       |<-(2)- tx_complete ----|       |
+        |       |--(3)- tx_add_output ->|       |
+        |       |<-(4)- tx_complete ----|       |
+        |       |--(5)- tx_add_input -->|       |
+        |   A   |<-(6)- tx_add_input ---|   B   |
+        |       |<-(7)- tx_add_output --|       |
+        |       |--(8)- tx_rm_output -->|       |
+        |       |<-(9)- tx_complete ----|       |
+        |       |--(10) tx_complete --->|       |
+        +-------+                       +-------+
+
+
+### The `tx_add_input` Message
+
+This message contains a transaction input.
+
+1. type: 66 (`tx_add_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+    * [`u16`:`prevtx_len`]
+    * [`prevtx_len*byte`:`prevtx`]
+    * [`u32`:`prevtx_vout`]
+    * [`u32`:`sequence`]
+    * [`u16`:`script_sig_len`]
+    * [`script_sig_len*byte`:`script_sig`]
+
+#### Requirements
+
+The sending node:
+  - MUST add all sent inputs to the transaction
+  - MUST use a unique `serial_id` for each input currently added to the
+    transaction
+  - MUST NOT re-transmit inputs it has received from the peer
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
+    - MUST send odd `serial_id`s
+
+The receiving node:
+  - MUST add all received inputs to the transaction
+  - MUST fail the negotiation if:
+    - the `prevtx` and `prevtx_vout` are identical to a previously added
+      (and not removed) input's
+    - `prevtx` is not a valid transaction
+    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
+    - the `prevtx_out` input of `prevtx` is not an `OP_0` to `OP_16`
+      followed by a single push
+    - the `serial_id` is already included in the transaction
+    - the `serial_id` has the wrong parity
+    - if has received 4096 `tx_add_input` messages during this negotiation
```

Comment:
If transaction size is superior to `MAX_STANDARD_TX_WEIGHT` MUST fail ?

Reply:
Is `MAX_STANDARD_TX_WEIGHT` a bitcoind constant?